File: LeetCode/src/problem001_100/Problem029.java
package problem001_100;

public class Problem029 {

    public int divide(int dividend, int divisor) {
        if (dividend == 0) {
            return 0;
        }

        if (dividend == divisor) {
            return 1;
        }

        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        if (divisor == Integer.MIN_VALUE) {
            return 0;
        }

        int originDividend = dividend;

        if (dividend == Integer.MIN_VALUE) {
            dividend += Math.abs(divisor);
        }

        boolean isNegative = false;

        if (dividend < 0 && divisor > 0 || dividend > 0 && divisor < 0) {
            isNegative = true;
        }

        dividend = Math.abs(dividend);
        divisor = Math.abs(divisor);

        int count = 0;

        while (divisor <= dividend) {
            int tmpDivisor = divisor;
            boolean isTooBigger = false;
            int tmpCount = 1;
            while (tmpDivisor <= dividend) {
                if ((tmpDivisor << 1) < 0) {
                    isTooBigger = true;
                    break;
                }
                tmpCount = tmpCount << 1;
                tmpDivisor = tmpDivisor << 1;
            }

            if (isTooBigger) {
                dividend -= tmpDivisor;
            } else {
                tmpCount = tmpCount >> 1;
                dividend -= (tmpDivisor >> 1);
            }

            count += tmpCount;
        }

        if (originDividend == Integer.MIN_VALUE) {
            count++;
        }

        if (isNegative) {
            return -count;
        }

        return count;
    }

    public static void main(String[] args) {
        System.out.println(new Problem029().divide(10, 3));
        System.out.println(new Problem029().divide(7, -3));
        System.out.println(new Problem029().divide(2, -3));
        System.out.println(new Problem029().divide(3, -3));
        System.out.println(new Problem029().divide(-2147483648, -1));
        System.out.println(new Problem029().divide(2147483647, 1));
        System.out.println(new Problem029().divide(20, 3));
        System.out.println(new Problem029().divide(-2147483648, 2));

    }
    
}

File: LeetCode/src/Problem029.kt
class Problem029 {

    fun divide(dividend: Int, divisor: Int): Int {
        if (divisor == 1) {
            return dividend
        }

        if (dividend == Int.MIN_VALUE && divisor == -1) {
            return Int.MAX_VALUE
        }

        val signDiff = (dividend > 0) xor (divisor > 0)
        var t = Math.abs(dividend.toLong())
        val d = Math.abs(divisor.toLong())

        var result = 0

        for (i in 31 downTo 0) {
            if ((t shr i) >= d) {
                result += 1 shl i
                t -= d shl i
            }
        }

        return if (signDiff) {
            -result
        } else {
            result
        }
    }

}

fun main(args: Array<String>) {
    val res = Problem029().divide(-2147483648, 2)
    println("res: $res")
}
